docs/public-async-nio-apis.md

```diff
@@ -0,0 +1,1162 @@
+    public func handlerRemoved(context: NIOCore.ChannelHandlerContext)
+
+    /// Called when a user inbound event has been triggered.
+    ///
+    /// This should call `context.fireUserInboundEventTriggered` to forward the operation to the next `_ChannelInboundHandler` in the `ChannelPipeline` if you want to allow the next handler to also handle the event.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    ///     - event: The event.
+    public func userInboundEventTriggered(context: NIOCore.ChannelHandlerContext, event: Any)
+
+    /// Called when some data has been read from the remote peer.
+    ///
+    /// This should call `context.fireChannelRead` to forward the operation to the next `_ChannelInboundHandler` in the `ChannelPipeline` if you want to allow the next handler to also handle the event.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    ///     - data: The data read from the remote peer, wrapped in a `NIOAny`.
+    public func channelRead(context: NIOCore.ChannelHandlerContext, data: NIOCore.NIOAny)
+
+    /// Called when the `Channel` has become inactive and is no longer able to send and receive data.
+    ///
+    /// This should call `context.fireChannelInactive` to forward the operation to the next `_ChannelInboundHandler` in the `ChannelPipeline` if you want to allow the next handler to also handle the event.
+    ///
+    /// - parameters:
+    ///     - context: The `ChannelHandlerContext` which this `ChannelHandler` belongs to.
+    public func channelInactive(context: NIOCore.ChannelHandlerContext)
+}
+```
+
+### HTTP/2.0
+
+```swift
+extension NIOHTTP2Handler {
+    /// A variant of `NIOHTTP2Handler.StreamMultiplexer` which creates a child channel for each HTTP/2 stream and
+    /// provides access to inbound HTTP/2 streams.
+    ///
+    /// In general in NIO applications it is helpful to consider each HTTP/2 stream as an
+    /// independent stream of HTTP/2 frames. This multiplexer achieves this by creating a
+    /// number of in-memory `HTTP2StreamChannel` objects, one for each stream. These operate
+    /// on ``HTTP2Frame/FramePayload`` objects as their base communication
+    /// atom, as opposed to the regular NIO `SelectableChannel` objects which use `ByteBuffer`
+    /// and `IOData`.
+    ///
+    /// Outbound stream channel objects are initialized upon creation using the supplied `streamStateInitializer` which returns a type
+    /// `Output`. This type may be `HTTP2Frame` or changed to any other type.
+    @available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+    public struct AsyncStreamMultiplexer<InboundStreamOutput> {
+        /// Create a stream channel initialized with the provided closure
+        public func createStreamChannel<Output: Sendable>(_ initializer: @escaping NIOChannelInitializerWithOutput<Output>) async throws -> Output
```

Comment:
I am still wondering if we should call this `createOutboundStream` this would provide a good balance to the `InboundStreamOutput` e.g. cc @rnro 
```suggestion
        public func createOutboundStream<Output: Sendable>(_ initializer: @escaping NIOChannelInitializerWithOutput<Output>) async throws -> Output
```

Reply:
If we are dropping consistency with the old multiplexer then I think `createOutboundStream` is an improvement, but I'm now actually wondering if we should just go for `createStream`, do we need to disambiguate against somehow creating an inbound stream? cc @glbrntt

Reply:
There's still something which feels a bit vague about how we call `createStream`/`createOutboundStream` but then hand back the caller basically anything. Our assumption is that they wrapped the channel in the initializer and returned that, which might be fair enough but I'm not sure that the word `initializer` conveys all that (new to async-land) responsibility.

Reply:
- `createStreamChannel` is slightly odd because we don't return a `Channel`
- `createOutboundStream` is better, but it's not possible to create inbound streams so I don't think the "outbound" distinction is important
- `createStream` is better still

> There's still something which feels a bit vague about how we call createStream/createOutboundStream but then hand back the caller basically anything. Our assumption is that they wrapped the channel in the initializer and returned that, which might be fair enough but I'm not sure that the word initializer conveys all that (new to async-land) responsibility.

This is true but I don't know if it's really an issue; regardless of what the user does we're still creating a stream.

Reply:
@FranzBusch I can see your point but I personally think it's okay because `createStream` requires that you pass in an initializer (maybe an argument for keeping that name) so it's clear that it won't use the `inboundStreamInitializer`.

Reply:
@glbrntt 
> regardless of what the user does we're still creating a stream

Good point

Reply:
We're assuming `inboundStreamIntializer` is the right name, which isn't necessarily the case (although it matters much less IMO because trailing closure syntax). Maybe it should also just be `initializer`.

Reply:
IMO naming it just `initializer` is a step back. We just had recent feedback in the forums that a users was confused about what that initializer does. So become more vague feels like a step in the wrong direction.

Reply:
> We just had recent feedback in the forums that a users was confused about what that initializer does. 

Can you share a link to this?

Neither `inboundStreamIntializer` nor `initializer` will tell you _what_ an initialiser does. `inboundStream` tells you what the initialiser applies to, without `inboundStream` you assume from the context what it applies to.

Reply:
The comment was in the Swift Open Source Slack actually and not in the forums. This was the users feedback

> update: i finally got an HTTP/2 request working, the piece i was missing was that you still need to add your application stream handler in NIOHTTP2Handler.StreamMultiplexer.createStreamChannel(_:), even if you already provided a callback when setting up the handlers with Channel.configureHTTP2Pipeline. which was very counterintuitive given how the docs are worded:
https://swiftinit.org/docs/swift-nio-http2/niohttp2/niohttp2handler.multiplexer
it gave the impression that if you provide the callback during outer setup then the multiplexer is “ready to go” and you can just return $0.eventLoop.createSucceededVoidFuture(). please update these docs :slightly_smiling_face:

To be fair it is more about the documentation explaining the different of the two closures and why both are needed.

Personally, I think there is still value in having outbound in the method name.

Reply:
Yeah that's absolutely about the docs and not about the naming. The linked docs do call out that it's an outbound stream which wasn't sufficient so I'm not sure keeping it in the name makes a difference. That said, this isn't a hill I'll die on!